include/boost/url/impl/url.hpp

```diff
@@ -22,6 +24,59 @@ segments(
     return urls::segments(*this, a);
 }
 
+template<class Charset>
+void
+url::
+normalize_octets_impl(
+    int id,
+    Charset const& cs) noexcept
```

Comment:
if this function is only ever used internally, consider changing `Charset const&` to `grammar::lut_chars const&`. This way, if the function is ever called with something that is not a `lut_chars`, we will be alerted with a compiler error. Otherwise, we would silently emit two copies of this function in the binary - and a goal of the library is to minimize the amount of object code.